From 6f58c6ca2a3e949fea8085776a22cb96f5b4a950 Mon Sep 17 00:00:00 2001
From: Randolph <randolph@andestech.com>
Date: Thu, 22 Feb 2024 20:15:11 +0800
Subject: [PATCH 168/191] Merge the Qilai PCIe from linux 5.4

Add the cpu address and pcie address translation, using the iATU.
Checking the link up state from the debug port by reading smlh & rdlh.
Add the sii base register by reading DTS.
Setup DMA addressing capabilities to 64bits.
Keep the forcedly link to PCIe Gen 1 code for future debug.

Signed-off-by: Randolph <randolph@andestech.com>
---
 drivers/pci/controller/dwc/Kconfig            |  9 ++++
 drivers/pci/controller/dwc/Makefile           |  1 +
 .../pci/controller/dwc/pcie-designware-host.c | 25 +++++++++++
 .../pci/controller/dwc/pcie-designware-plat.c |  5 +++
 drivers/pci/controller/dwc/pcie-designware.c  | 13 +++++-
 drivers/pci/controller/dwc/pcie-designware.h  |  1 +
 drivers/pci/controller/dwc/pcie-qilai.c       | 24 +++++++++++
 drivers/pci/controller/dwc/pcie-qilai.h       | 42 +++++++++++++++++++
 8 files changed, 119 insertions(+), 1 deletion(-)
 create mode 100644 drivers/pci/controller/dwc/pcie-qilai.c
 create mode 100644 drivers/pci/controller/dwc/pcie-qilai.h

diff --git a/drivers/pci/controller/dwc/Kconfig b/drivers/pci/controller/dwc/Kconfig
index ae9d083c406f..b85ff4575848 100644
--- a/drivers/pci/controller/dwc/Kconfig
+++ b/drivers/pci/controller/dwc/Kconfig
@@ -188,6 +188,15 @@ config PCIE_QCOM_EP
 	  to work in endpoint mode. The PCIe controller uses the DesignWare core
 	  plus Qualcomm-specific hardware wrappers.
 
+config PCIE_QILAI
+	bool "Andes Qilai PCIe host controller"
+	depends on PCI_MSI_IRQ_DOMAIN
+	depends on SOC_QILAI || COMPILE_TEST
+	select PCIE_DW_HOST
+	help
+	  Say Y here if you want PCIe controller support for the Andes Qilai
+	  Platform.
+
 config PCIE_ARMADA_8K
 	bool "Marvell Armada-8K PCIe controller"
 	depends on ARCH_MVEBU || COMPILE_TEST
diff --git a/drivers/pci/controller/dwc/Makefile b/drivers/pci/controller/dwc/Makefile
index 8ba7b67f5e50..dcf72fa223ef 100644
--- a/drivers/pci/controller/dwc/Makefile
+++ b/drivers/pci/controller/dwc/Makefile
@@ -13,6 +13,7 @@ obj-$(CONFIG_PCI_LAYERSCAPE) += pci-layerscape.o
 obj-$(CONFIG_PCI_LAYERSCAPE_EP) += pci-layerscape-ep.o
 obj-$(CONFIG_PCIE_QCOM) += pcie-qcom.o
 obj-$(CONFIG_PCIE_QCOM_EP) += pcie-qcom-ep.o
+obj-$(CONFIG_PCIE_QILAI) += pcie-qilai.o
 obj-$(CONFIG_PCIE_ARMADA_8K) += pcie-armada8k.o
 obj-$(CONFIG_PCIE_ARTPEC6) += pcie-artpec6.o
 obj-$(CONFIG_PCIE_ROCKCHIP_DW_HOST) += pcie-dw-rockchip.o
diff --git a/drivers/pci/controller/dwc/pcie-designware-host.c b/drivers/pci/controller/dwc/pcie-designware-host.c
index 39f3b37d4033..69ded77203ae 100644
--- a/drivers/pci/controller/dwc/pcie-designware-host.c
+++ b/drivers/pci/controller/dwc/pcie-designware-host.c
@@ -18,6 +18,7 @@
 
 #include "../../pci.h"
 #include "pcie-designware.h"
+#include "pcie-qilai.h"
 
 static struct pci_ops dw_pcie_ops;
 static struct pci_ops dw_child_pcie_ops;
@@ -367,9 +368,11 @@ static int dw_pcie_msi_host_init(struct dw_pcie_rp *pp)
 						    dw_chained_msi_isr, pp);
 	}
 
+#ifndef CONFIG_SOC_QILAI
 	ret = dma_set_mask_and_coherent(dev, DMA_BIT_MASK(32));
 	if (ret)
 		dev_warn(dev, "Failed to set DMA mask to 32-bit. Devices with only 32-bit MSI support may not work properly\n");
+#endif
 
 	msi_vaddr = dmam_alloc_coherent(dev, sizeof(u64), &pp->msi_data,
 					GFP_KERNEL);
@@ -392,6 +395,7 @@ int dw_pcie_host_init(struct dw_pcie_rp *pp)
 	struct pci_host_bridge *bridge;
 	struct resource *res;
 	int ret;
+	int val;
 
 	raw_spin_lock_init(&pp->lock);
 
@@ -415,6 +419,20 @@ int dw_pcie_host_init(struct dw_pcie_rp *pp)
 			return PTR_ERR(pci->dbi_base);
 	}
 
+#ifdef CONFIG_SOC_QILAI
+	if (!pci->sii_base) {
+		res = platform_get_resource_byname(pdev, IORESOURCE_MEM, "sii");
+		pci->sii_base = devm_ioremap_resource(dev, res);
+		if (IS_ERR(pci->sii_base))
+			return PTR_ERR(pci->sii_base);
+
+		/* Enable msi */
+		dw_pcie_read(pci->sii_base + PCIE_CONTROLLER_INTERRUPT_CONTROL1, 0x4, &val);
+		val |= QILAI_MSI_CTRL_INT_EN;
+		dw_pcie_write(pci->sii_base + PCIE_CONTROLLER_INTERRUPT_CONTROL1, 0x4, val);
+	}
+#endif
+
 	bridge = devm_pci_alloc_host_bridge(dev, 0);
 	if (!bridge)
 		return -ENOMEM;
@@ -484,6 +502,13 @@ int dw_pcie_host_init(struct dw_pcie_rp *pp)
 			goto err_free_msi;
 	}
 
+#ifdef CONFIG_SOC_QILAI
+	/* Do phy link up */
+	dw_pcie_read(pci->sii_base + PCIE_GEN_CONTROL2, 0x4, &val);
+	val |= CFG_LTSSM_EN;
+	dw_pcie_write(pci->sii_base + PCIE_GEN_CONTROL2, 0x4, val);
+#endif
+
 	/* Ignore errors, the link may come up later */
 	dw_pcie_wait_for_link(pci);
 
diff --git a/drivers/pci/controller/dwc/pcie-designware-plat.c b/drivers/pci/controller/dwc/pcie-designware-plat.c
index 1fcfb840f238..a20266957bb4 100644
--- a/drivers/pci/controller/dwc/pcie-designware-plat.c
+++ b/drivers/pci/controller/dwc/pcie-designware-plat.c
@@ -133,6 +133,11 @@ static int dw_plat_pcie_probe(struct platform_device *pdev)
 
 	platform_set_drvdata(pdev, dw_plat_pcie);
 
+#ifdef CONFIG_SOC_QILAI
+	if (dma_set_mask_and_coherent(dev, DMA_BIT_MASK(64)))
+		dev_warn(dev, "No suitable DMA available\n");
+#endif
+
 	switch (dw_plat_pcie->mode) {
 	case DW_PCIE_RC_TYPE:
 		if (!IS_ENABLED(CONFIG_PCIE_DW_PLAT_HOST))
diff --git a/drivers/pci/controller/dwc/pcie-designware.c b/drivers/pci/controller/dwc/pcie-designware.c
index 575834cae3b9..2d973b301322 100644
--- a/drivers/pci/controller/dwc/pcie-designware.c
+++ b/drivers/pci/controller/dwc/pcie-designware.c
@@ -18,6 +18,7 @@
 
 #include "../../pci.h"
 #include "pcie-designware.h"
+#include "pcie-qilai.h"
 
 void dw_pcie_version_detect(struct dw_pcie *pci)
 {
@@ -312,7 +313,7 @@ static int __dw_pcie_prog_outbound_atu(struct dw_pcie *pci, u8 func_no,
 {
 	u32 retries, val;
 	u64 limit_addr;
-
+#ifndef CONFIG_SOC_QILAI
 	if (pci->ops && pci->ops->cpu_addr_fixup)
 		cpu_addr = pci->ops->cpu_addr_fixup(pci, cpu_addr);
 
@@ -323,6 +324,10 @@ static int __dw_pcie_prog_outbound_atu(struct dw_pcie *pci, u8 func_no,
 	    !IS_ALIGNED(pci_addr, pci->region_align) || !size) {
 		return -EINVAL;
 	}
+#else
+	cpu_addr = qilai_pcie_cpu_addr_fixup(pci, cpu_addr);
+	limit_addr = cpu_addr + size - 1;
+#endif
 
 	dw_pcie_writel_atu_ob(pci, index, PCIE_ATU_LOWER_BASE,
 			      lower_32_bits(cpu_addr));
@@ -470,9 +475,15 @@ int dw_pcie_link_up(struct dw_pcie *pci)
 	if (pci->ops && pci->ops->link_up)
 		return pci->ops->link_up(pci);
 
+#ifndef CONFIG_SOC_QILAI
 	val = dw_pcie_readl_dbi(pci, PCIE_PORT_DEBUG1);
 	return ((val & PCIE_PORT_DEBUG1_LINK_UP) &&
 		(!(val & PCIE_PORT_DEBUG1_LINK_IN_TRAINING)));
+#else
+	/* Read smlh & rdlh link up by checking debug port */
+	dw_pcie_read(pci->sii_base + LINK_DEBUG_DATA, 0x4, &val);
+	return ((val & SMLH_LINK_UP) && (val & RDLH_LINK_UP));
+#endif
 }
 EXPORT_SYMBOL_GPL(dw_pcie_link_up);
 
diff --git a/drivers/pci/controller/dwc/pcie-designware.h b/drivers/pci/controller/dwc/pcie-designware.h
index a871ae7eb59e..177cc20dbe91 100644
--- a/drivers/pci/controller/dwc/pcie-designware.h
+++ b/drivers/pci/controller/dwc/pcie-designware.h
@@ -307,6 +307,7 @@ struct dw_pcie {
 	void __iomem		*dbi_base;
 	void __iomem		*dbi_base2;
 	void __iomem		*atu_base;
+	void __iomem		*sii_base;
 	size_t			atu_size;
 	u32			num_ib_windows;
 	u32			num_ob_windows;
diff --git a/drivers/pci/controller/dwc/pcie-qilai.c b/drivers/pci/controller/dwc/pcie-qilai.c
new file mode 100644
index 000000000000..63ee167c5c60
--- /dev/null
+++ b/drivers/pci/controller/dwc/pcie-qilai.c
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Driver for the Andes ATCDMAC300
+ *
+ * Copyright (C) 2021 Andes Technology Corporation
+ *
+ */
+#include "pcie-qilai.h"
+
+
+u64 qilai_pcie_cpu_addr_fixup(struct dw_pcie *pcie, u64 cpu_addr)
+{
+	if ((cpu_addr >= DM1_EP_MEM_MAP_BASE) &&
+	    (cpu_addr < DM1_EP_MEM_MAP_END)) {
+		return cpu_addr - DM1_EP_MEM_MAP_BASE;
+	} else if ((cpu_addr >= DM0_EP_MEM_MAP_BASE) &&
+		   (cpu_addr < DM0_EP_MEM_MAP_END)) {
+		return cpu_addr - DM0_EP_MEM_MAP_BASE;
+	} else if ((cpu_addr >= DM2_EP_MEM_MAP_BASE) &&
+		   (cpu_addr < DM2_EP_MEM_MAP_END)) {
+		return cpu_addr - DM2_EP_MEM_MAP_BASE;
+	}
+	return -EINVAL;
+}
\ No newline at end of file
diff --git a/drivers/pci/controller/dwc/pcie-qilai.h b/drivers/pci/controller/dwc/pcie-qilai.h
new file mode 100644
index 000000000000..4e1ccf48ac06
--- /dev/null
+++ b/drivers/pci/controller/dwc/pcie-qilai.h
@@ -0,0 +1,42 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ * Header file for the Andes Qilai PCIe Controller driver
+ *
+ * Copyright (C) 2024 Andes Technology Corporation
+ */
+#ifndef PCIE_QILAI_H
+#define	PCIE_QILAI_H
+
+#include <linux/bitfield.h>
+#include <linux/dma-mapping.h>
+#include <linux/irq.h>
+#include <linux/msi.h>
+#include <linux/pci.h>
+
+#include "pcie-designware.h"
+
+#define PCIE_CONTROLLER_INTERRUPT_CONTROL1        0x15C
+#define QILAI_MSI_CTRL_INT_EN                     BIT(28)
+#define PF0_PORT_LOGIC_COHERENCY_CONTROL_3_OFF    0x8E8
+#define CFG_MSTR_ARCACHE_MODE                     GENMASK(6, 3)
+#define CFG_MSTR_AWCACHE_MODE                     GENMASK(14, 11)
+#define CFG_MSTR_ARCACHE_VALUE                    GENMASK(22, 19)
+#define CFG_MSTR_AWCACHE_VALUE                    GENMASK(30, 27)
+
+#define PCIE_GEN_CONTROL2		0x54
+#define CFG_LTSSM_EN			BIT(0)
+
+#define LINK_DEBUG_DATA             0xC0
+#define SMLH_LINK_UP                BIT(6)
+#define RDLH_LINK_UP                BIT(7)
+
+#define DM0_EP_MEM_MAP_BASE	0x2000000000
+#define DM0_EP_MEM_MAP_END	0x4000000000
+#define DM1_EP_MEM_MAP_BASE	0x1000000000
+#define DM1_EP_MEM_MAP_END	0x1800000000
+#define DM2_EP_MEM_MAP_BASE	0x1800000000
+#define DM2_EP_MEM_MAP_END	0x2000000000
+
+u64 qilai_pcie_cpu_addr_fixup(struct dw_pcie *pcie, u64 cpu_addr);
+
+#endif
\ No newline at end of file
-- 
2.34.1

